### PR TITLE
feat: Add tests and implementation of kmp string algorithm

### PR DESCRIPTION
**Pull request by ArielXL**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,6 +13,8 @@ Here is a checklist of the Data Structures and Algorithms we have or we are plan
 - [x] [Union-Find](./data_structures/union_find/implementation.py)
 - [x] [Graphs](./data_structures/graphs/implementation.py)
 - [ ] Linked List
+- [ ] Disjoint Set
```
Comment (albexl) [inline README.md:16]: Two things to notice here:
- We already have the Disjoint Sets data structure added to our list. It is under the name of Union-Find. So it is not necessary to add it.
- Also, it would be great if this PR is only focused on what it says in the title: Adding KMP implementation and tests for that implementation. Maybe add new items to the To-Do list in another PR.

File: algorithms/searching/kmp.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+'''
```
Comment (albexl) [inline algorithms/searching/kmp.py:1]: This looks great. Can you add the naive implementation in another PR following this same Class hierarchy?

File: tests/test_kmp.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+"""Module to test the KMP implementations."""
+
+from unittest import TestCase
+from algorithms.searching.kmp import KMP
+
+
+class TestKMP(TestCase):
```
Comment (albexl) [inline tests/test_kmp.py:7]: Can you try adding the test cases using the `parameterized` library like it is being used [here](https://github.com/albexl/data-structures-for-teaching/blob/dev/tests/test_sorting.py#L19)? Comment if you have any questions on how to do it.

**Comment by albexl** [comment]
@ArielXL while we review the PR, can you add an issue and link this PR? So others know that this is in progress.

File: README.md (hunk 1 above)
Comment (ArielXL) [reply README.md:16]: ok, my mistake

File: algorithms/searching/kmp.py (hunk 2 above)
Comment (ArielXL) [reply algorithms/searching/kmp.py:1]: sure my friend

**Comment by albexl** [comment]
Also, it would be great if you could check the `check item` corresponding to the `KMP` on the [README.md](https://github.com/albexl/data-structures-for-teaching/blob/dev/README.md) file once you have completed this implementation. This way it is marked as `implemented` and people can try different algorithms or data structures.

File: tests/test_kmp.py (hunk 3 above)
Comment (ArielXL) [reply tests/test_kmp.py:7]: i'm going to look at the library

**Comment by albexl** [comment]
@ArielXL I added some commits to try to fix the comments I left yesterday. Please review them and let me know what you think. If everything looks fine for you, we can merge it.

**Comment by ArielXL** [comment]
@albexl why did you remove the segment tree data structure?

**Comment by ArielXL** [comment]
the next I'm going to do is the naive algorithm, do you think?

**Comment by albexl** [comment]
> @albexl why did you remove the segment tree data structure?

@ArielXL  Because I think it should be part of another PR that just adds some more items to the To-Do list. This should be focused on adding KMP and the tests, as the title says. I think I mentioned this in my first comment above.

**Comment by albexl** [comment]
@ArielXL can we merge this then? Looks great to me. Let me know if you have any suggestions for improvement.

**Comment by ArielXL** [comment]
@al

> @ArielXL can we merge this then? Looks great to me. Let me know if you have any suggestions for improvement.

yep
